docs: add cryptographic docs (#2278)
canonical/charmcraft: docs/explanation/cryptography.rst

.. _explanation_cryptographic-technology:

Cryptographic technology in Charmcraft
======================================

Charmcraft uses cryptographic technologies to fetch arbitrary files over the internet,
communicate with local processes, and store credentials. It does not directly implement
its own cryptography, but it does depend on external libraries to do so.

> Charmcraft is built upon `Craft Application`_ and derives much of its functionality from

Reviewer: ```suggestion
Charmcraft is built on `Craft Application`_ and derives much of its functionality from
```

I wish I hadn't let this word propagate...

Author: "We built this city upon rock and roll." -- Starcraft

it, so much of Charmcraft's cryptographic functionality is described in `Cryptographic
technology in Craft Application`_. The additional cryptography functionality in
Charmcraft is documented below.

Local Docker images
~~~~~~~~~~~~~~~~~~~

> Charmcraft uses the `Docker SDK for Python`_ to communicate with a local Docker instance
> (if one is installed or running). This communicates with Docker over a Unix-domain
> socket if the user running Charmcraft has the necessary permission.

Reviewer: ```suggestion
If a Docker instance is installed or running on the host, Charmcraft uses the `Docker
SDK for Python`_ to communicate with it. The SDK communicates with Docker over a
Unix-domain socket if the user running Charmcraft has the necessary permission.
```

> socket if the user running Charmcraft has the necessary permission.

Reviewer: > installed or running on the host

I would rather we explain what the container could be doing and when Charmcraft would need to communicate with it.

Reviewer: Sublime!
